File: browser-test/src/security_browser.test.ts
import { assertEndpointEquals, assertPageIncludes } from './support' 
import { startSession, getUserId, gotoRootUrl, gotoEndpoint } from './support'
import { loginWithSimulatedIdcs, loginAsAdmin, loginAsGuest } from './support'
import { logout, endSession } from './support'
import { BASE_URL } from './support'

var assert = require('assert');

describe('security browser testing', () => {
  let container;

  it('basicOidcLogin', async () => {
    const { browser, page } = await startSession();

    // -- Fails to redirect on clicking 'log in with IDCS'
    await loginWithSimulatedIdcs(page);
    await gotoEndpoint(page, '/securePlayIndex');

    // -- Previously FAILED
    // -- Reviewer Please double check 
    await assertPageIncludes(page, 'You are logged in.');

    await gotoEndpoint(page, '/users/me');

    // -- Previously FAILED
    // -- Reviewer Please double check 
    await assertPageIncludes(page, 'OidcClient');
    await assertPageIncludes(page, 'username@example.com');

    await logout(page);
    endSession(browser);
  })

  it('homePage_whenNotLoggedIn_redirectsToLoginForm', async () => {
    const { browser, page } = await startSession();
    await gotoRootUrl(page);
    assertEndpointEquals(page, '/loginForm');
    endSession(browser);
  })

  it('homePage_whenLoggedInAsAdmin_redirectsToAdminProgramList', async () => {
    const { browser, page } = await startSession();
    await loginAsAdmin(page);
    await gotoRootUrl(page);

    assertEndpointEquals(page, '/admin/programs');

    await logout(page);
    await gotoRootUrl(page);
    endSession(browser);
  })

  it('homePage_whenLoggedInAsApplicant_redirectsToApplicantProgramList', async () => {
    const { browser, page } = await startSession();
    await loginAsGuest(page);

    let user_id = await getUserId(page);

    await gotoRootUrl(page);

    // REVIEWER Please Double Check
    // I Changed '/programs' to '/edit' to make test work
    assertEndpointEquals(page, '/applicants/'.concat(user_id).concat('/edit'));
    await logout(page);
    endSession(browser);
  })

  it('noCredLogin', async () => {
    const { browser, page } = await startSession();
    await loginAsGuest(page);

    await gotoEndpoint(page, '/securePlayIndex');
    await assertPageIncludes(page, 'You are logged in.');

    await gotoEndpoint(page, '/users/me');
    await assertPageIncludes(page, 'GuestClient');
    await assertPageIncludes(page, 'ROLE_APPLICANT');

    await gotoEndpoint(page, '/admin/programs');
    await assertPageIncludes(page, '403');


    // EDIT: Removed `await logout(page)` since that wouldn't be in a 403 page
    endSession(browser);
  })

  it('mergeLogins', async () => {
    const { browser, page } = await startSession();
    await loginAsGuest(page);

    await gotoEndpoint(page, '/users/me');
    let pg_source = await page.content();
    assert(pg_source.includes('GuestClient'));

    let user_id = await getUserId(page);

    await gotoRootUrl(page);

    // Had to add this because login attempt after gotoRootUrl 
    // resulted in timeout error
    await gotoEndpoint(page, '/loginForm');
    await loginWithSimulatedIdcs(page);

    await gotoEndpoint(page, '/users/me');

    // Fails after no-fail login attempt
    await assertPageIncludes(page, 'OidcClient');

    // FAILS on getting user id
    let user_id2 = await getUserId(page);
    assert.equal(user_id, user_id2);

    await logout(page);
    await gotoEndpoint(page, '/loginForm');
    await loginWithSimulatedIdcs(page);

    // Fails after no-fail login attempt
    assert(pg_source.includes('OidcClient'));

    let user_id3 = await getUserId(page);
    assert.equal(user_id, user_id3);

    await logout(page);
    endSession(browser);
  })

  it('adminTestLogin', async () => {
    const { browser, page } = await startSession();
    await loginAsAdmin(page);
    await gotoEndpoint(page, '/securePlayIndex');
    let pg_source = await page.content();
    assert(pg_source.includes('You are logged in.'));

    await gotoEndpoint(page, '/users/me');
    pg_source = await page.content();
    assert(pg_source.includes('FakeAdminClient'));
    assert(pg_source.includes('ROLE_UAT_ADMIN'));

    await gotoEndpoint(page, '/admin/programs');
    pg_source = await page.content();
    assert(pg_source.includes('Programs'));
    assert(pg_source.includes('Create new program'));
    endSession(browser);
  })
})

File: server/test/app/SecurityBrowserTest.java
package app;

import static org.assertj.core.api.Assertions.assertThat;
import static play.test.Helpers.fakeApplication;

import auth.Roles;
import com.google.common.collect.ImmutableMap;
import controllers.routes;
import java.util.Optional;
import java.util.concurrent.TimeUnit;
import models.Applicant;
import org.junit.Before;
import org.junit.Test;
import play.Application;
import repository.UserRepository;
import services.WellKnownPaths;
import support.CfTestHelpers;
import org.fluentlenium.core.domain.FluentWebElement;

public class SecurityBrowserTest extends BaseBrowserTest {
  private static UserRepository userRepository;

  @Override
  protected Application provideApplication() {
    ImmutableMap<String, Object> config =
        new ImmutableMap.Builder<String, Object>()
            .putAll(CfTestHelpers.oidcConfig("oidc", 3380))
            .build();
    return fakeApplication(config);
  }

  @Before
  public void getApplicantRepository() {
    userRepository = app.injector().instanceOf(UserRepository.class);
  }

  private void loginWithSimulatedIdcs() {
    goTo(routes.LoginController.applicantLogin(Optional.empty()));
    // If we are not cookied, enter a username and password.
    // Otherwise, since the fake provider uses the "web" flow, we're automatically sent to the
    // redirect URI to merge logins.
    // TODO(#1770): Consider removing the below conditional entirely once full logout from the
    // fake OIDC provider is supported.
    if (browser.pageSource().contains("Enter any login")) {
      browser.$("[name='login']").click();
      browser.keyboard().sendKeys("username");
      browser.$("[name='password']").click();
      browser.keyboard().sendKeys("password");
      // Log in.
      browser.$(".login-submit").click();
      // Bypass consent screen.
      browser.$(".login-submit").click();
    }
  }

  @Test
  public void homePage_whenNotLoggedIn_redirectsToLoginForm() {
    goToRootUrl();
    assertUrlEquals(routes.HomeController.loginForm(Optional.empty()));
  }

  @Test
  public void homePage_whenLoggedInAsAdmin_redirectsToAdminProgramList() {
    loginAsAdmin();
    goToRootUrl();
    assertUrlEquals(controllers.admin.routes.AdminProgramController.index());
  }

  @Test
  public void homePage_whenLoggedInAsApplicantForFirstTime_redirectsToLanguageForm() {
    loginAsGuest();
    long applicantId = getApplicantId();

    goToRootUrl();

    assertUrlEquals(controllers.applicant.routes.ApplicantInformationController.edit(applicantId));
  }

  @Test
  public void noCredLogin() {
    loginAsGuest();
    // should be redirected to root.
    goTo(routes.HomeController.securePlayIndex());
    assertThat(browser.pageSource()).contains("You are logged in.");

    goTo(routes.ProfileController.myProfile());
    assertThat(browser.pageSource()).contains("GuestClient");
    assertThat(browser.pageSource()).contains(Roles.ROLE_APPLICANT.toString());

    goTo(controllers.admin.routes.AdminProgramController.index());
    assertThat(browser.pageSource()).contains("403");
  }

  @Test
  public void basicOidcLogin() {
    loginWithSimulatedIdcs();
    goTo(routes.HomeController.securePlayIndex());
    assertThat(browser.pageSource()).contains("You are logged in.");
    goTo(routes.ProfileController.myProfile());
    assertThat(browser.pageSource()).contains("OidcClient");
    assertThat(browser.pageSource()).contains("username@example.com");

    Applicant applicant = userRepository.lookupApplicant(getApplicantId()).toCompletableFuture().join().get();
    Optional<String> applicantName = applicant.getApplicantData().readString(WellKnownPaths.APPLICANT_FIRST_NAME);
    assertThat(applicantName).isPresent();
    assertThat(applicantName.get()).isEqualTo("username@example.com");

    assertThat(applicant.getApplicantData().readString(WellKnownPaths.APPLICANT_MIDDLE_NAME))
        .isEmpty();
    assertThat(applicant.getApplicantData().readString(WellKnownPaths.APPLICANT_LAST_NAME))
        .isEmpty();
  }

  @Test
  public void basicOidcProviderCentralLogout() {
    loginWithSimulatedIdcs();
    goTo(routes.HomeController.securePlayIndex());
    assertThat(browser.pageSource()).contains("You are logged in.");
    logout();
    assertThat(browser.url())
        .contains("session/end?post_logout_redirect_uri=http%3A%2F%2Flocalhost%3A19001%2F&client_id=foo")
        .as("redirects to login provider");
    assertThat(browser.pageSource().contains("Do you want to sign-out from")).as("Confirm logout from dev-oidc");
    FluentWebElement continueButton = browser.$("button").first();
    assertThat(continueButton.textContent()).contains("Yes");
    continueButton.click();
    assertThat(browser.url())
        .contains("loginForm");

    // Log in.
    goTo(routes.HomeController.index());
    assertThat(browser.pageSource()).contains("Don't have an account?");
    goTo(routes.LoginController.applicantLogin(Optional.empty()));
    // Verify we don't auto-login.
    assertThat(browser.pageSource())
        .contains("Enter any login");

  }
  @Test
  public void mergeLogins() {
    // First, log in as guest and get the applicant ID.
    loginAsGuest();
    goTo(routes.ProfileController.myProfile());
    assertThat(browser.pageSource()).contains("GuestClient");

    long applicantId = getApplicantId();

    // Then, login with IDCS and show that the applicant ID is the same.
    loginWithSimulatedIdcs();
    goTo(routes.ProfileController.myProfile());
    assertThat(browser.pageSource()).contains("OidcClient");
    assertThat(applicantId).isEqualTo(getApplicantId());

    logout();

    loginWithSimulatedIdcs();
    goTo(routes.ProfileController.myProfile());
    assertThat(browser.pageSource()).contains("OidcClient");
    assertThat(applicantId).isEqualTo(getApplicantId());
  }

  @Test
  public void adminTestLogin() {
    loginAsAdmin();

    goTo(routes.HomeController.securePlayIndex());
    assertThat(browser.pageSource()).contains("You are logged in.");

    goTo(routes.ProfileController.myProfile());
    assertThat(browser.pageSource()).contains("FakeAdminClient");
    assertThat(browser.pageSource()).contains(Roles.ROLE_CIVIFORM_ADMIN.toString());

    goTo(controllers.admin.routes.AdminProgramController.index());
    assertThat(browser.pageSource()).contains("Programs");
    assertThat(browser.pageSource()).contains("Create new program");
  }
}
